python/src/eval.cc

```diff
@@ -0,0 +1,72 @@
+#include "internal/eval.hh"
+#include "internal/errors.hh"
+#include "internal/nix-to-python.hh"
+#include "internal/python-to-nix.hh"
+#include <Python.h>
+
+#include <cxxabi.h>
+#include <store-api.hh>
+#include <shared.hh>
+
+namespace pythonnix {
+
+const char *currentExceptionTypeName() {
+  int status;
+  auto res = abi::__cxa_demangle(abi::__cxa_current_exception_type()->name(), 0,
+                                 0, &status);
+  return res ? res : "(null)";
+}
+
+static PyObject *_eval(const char *expression, PyObject *vars) {
+  nix::Strings storePath;
+  nix::EvalState state(storePath, nix::openStore());
+
+  nix::Env *env = nullptr;
+  auto staticEnv = pythonToNixEnv(state, vars, &env);
+  if (!staticEnv) {
+    return nullptr;
+  }
+  auto staticEnvPointer = std::make_shared<nix::StaticEnv>(*staticEnv);
+
+  auto e = state.parseExprFromString(expression, ".", staticEnvPointer);
+  nix::Value v;
+  e->eval(state, *env, v);
+
+  state.forceValueDeep(v);
+
+  nix::PathSet context;
+  return nixToPythonObject(state, v, context);
+}
+
+PyObject *eval(PyObject *self, PyObject *args, PyObject *keywds) {
+  const char *expression = nullptr;
+  PyObject *vars = nullptr;
+
+  // TODO: Only do this once
+  // By default, Nix sets the build-hook to be "$(readlink /proc/self/exe) __build-remote", expecting the current binary to be Nix itself.
+  // But when we call the Nix library from Python this isn't the case, the current binary is Python then
+  // So we need to change this default, pointing it to the Nix binary instead
+  // Because we can't know the path to the Nix binary from here, we use a preprocessor macro,
+  // which we can then influence from Meson, which we can then influence from the Nix build definition
+  nix::settings.buildHook = NIX_BUILD_HOOK;
+  // And by setting buildHook before calling initNix, we can override the defaults without overriding the user-provided options from the config files
+  nix::initNix();
```

Comment:
@roberth I am not yet fully sure how this relates to https://github.com/NixOS/nix/pull/7732, but this might interest you

Reply:
`initNix()` can only be called once. You can use `std::call_once` or the static local initializer trick, e.g.
```
    [[maybe_unused]] static auto init = [&]() -> bool
    {
        initNix();
        return true;
    }();
```

Reply:
Notice that nix also installs signal handler which will not play well with Python ones. I disabled those in my project like this: https://github.com/nix-community/harmonia/commit/d8af9885183ffecde2c677a57c0bf3a890ff7d20

Reply:
It's now called only once by using the python module initialization code that was already used for `initGC`: https://github.com/NixOS/nix/blob/6becd3e8945b842c5e4c657d7800e985ea9af08a/python/src/python-module.cc#L30-L36